### ISSUE TITLE
Excessive Sudden Resource Usage

### ISSUE THREAD

**Issue by Techeryy** [coder/coder#12085]
I had previously been running Coder v2.7.1 with no issues, however following a recent update to v2.7.2 I have ran into a strange behaviour, where if at least one workspace is running with a user connected the whole server will periodically freeze, reporting 100% CPU & 100% RAM utilisation. What makes this behaviour even more strange is that a user may be idle in their workspace or just writing code without even running it when this happens. Sometimes the server will recover from this after 30-60 seconds but other times it won't un-freeze (even after being left for 2+ hours) and require a reboot. After such a reboot Coder will fail to load workspaces requiring it to be redeployed to restore functionality. 

Typically with a couple of users connected to their workspaces the server will sit at about 40-60% CPU & 55-75% RAM, and the maximum number of users ever online at a time is two.

I then redeployed v2.7.1 for a period of time and there was still no issues, however when running 2.7.2 or 2.8.0 the issue mentioned above occurs relatively often.

I am running Coder within a docker container on my Ubuntu VPS, workspaces use a modified version of the docker template.
OS: Ubuntu 22.04.3 LTS
Current Coder Version: v2.8.1+908d236

**Comment by ericpaulsen** [comment]
hi @Techeryy - can you confirm whether Coder is running on a VM or Kubernetes, and how much CPU and RAM is allocated to the server?

**Comment by Techeryy** [comment]
> hi @Techeryy - can you confirm whether Coder is running on a VM or Kubernetes, and how much CPU and RAM is allocated to the server?

I am running Coder within docker with workspaces also provisioned by docker. I installed Coder using docker compose by following the steps on the official coder Docs, see https://coder.com/docs/v2/latest/install/docker for the instructions I followed.

Currently 2GB of RAM & 2 Cores are allocated to Coder, since generally it’s only me using it (for remote work) but occasionally one other user may be online. However I wouldn’t have thought this behaviour would have been caused by a low amount of resources since it doesn’t seem to correlate with the activity in the workspace, and it also came about in a small update.

**Comment by johnstcn** [comment]
> the whole server will periodically freeze, reporting 100% CPU & 100% RAM utilisation
To clarify, you mean the Coder container itself?
Is the ramp-up to 100% CPU/Memory gradual, or does it happen suddenly?

> After such a reboot Coder will fail to load workspaces requiring it to be redeployed to restore functionality.

Can you clarify what "redeployed" means in this case?

**Comment by Techeryy** [comment]
> > the whole server will periodically freeze, reporting 100% CPU & 100% RAM utilisation
> 
> To clarify, you mean the Coder container itself? Is the ramp-up to 100% CPU/Memory gradual, or does it happen suddenly?
> 
> > After such a reboot Coder will fail to load workspaces requiring it to be redeployed to restore functionality.
> 
> Can you clarify what "redeployed" means in this case?

Coder is running within in a docker container on it's own VPS with 2 Cores & 2GB of RAM, I haven't setup any sort of limit on how much of the available recourses a given container is allowed to use. When the behaviour occurs the control panel displays that the CPU & RAM usage of the entire VPS hits 100%, this happens in a sudden spike, and will remain at a solid 100% for 30-60 seconds or sometimes it will never return to normal levels. Now this would usually indicate to me that it requires more recourses however since it was working perfectly before a minor bug fix & that the behaviour occurs seemingly randomly when workspaces are running I feel like there is some sort of bug.

If the VPS remained at 100% I restart the entire VPS (which obviously restarts all containers), however this usually bricks the coder install. So I copy all of my workspace files to elsewhere in the filesystem and remove Coder from the machine (including networks, containers, volumes etc), I then re-pull the Coder image using docker compose and completely set everything back up, copy my files over & then I have a working install which works until the next time it happens.

Hope that clears up any confusion, don't hesitate to ask if you need more info.

**Comment by johnstcn** [comment]
How do your users normally interact with Coder? Over SSH, or via browser applications (for example, code-server)?

**Comment by coadler** [comment]
One thing you can try setting is `GOMAXPROCS=1` in the container env, so Coder doesn't lock up the whole server. This will force Coder to only use 1 core. To help figure out where the CPU is going, you can add `CODER_PPROF_ENABLE=true` and forward port `6060` to the host.

When you see the CPU pinning, run `curl -o cpu-trace.pprof http://localhost:6060/debug/pprof/trace?seconds=10` to grab a CPU profile. If you could send it back to us via github or email (`<first name>@coder.com` for any of us), that would be great.

**Comment by Techeryy** [comment]
> How do your users normally interact with Coder? Over SSH, or via browser applications (for example, code-server)?

All user interaction is via the code-server browser application.

> One thing you can try setting is `GOMAXPROCS=1` in the container env, so Coder doesn't lock up the whole server. This will force Coder to only use 1 core. To help figure out where the CPU is going, you can add `CODER_PPROF_ENABLE=true` and forward port `6060` to the host.
> 
> When you see the CPU pinning, run `curl -o cpu-trace.pprof http://localhost:6060/debug/pprof/trace?seconds=10` to grab a CPU profile. If you could send it back to us via github or email (`<first name>@coder.com` for any of us), that would be great.

I will look into generating some logs using the steps you specified tomorrow & then post them here once I have done it!
Thank You

**Comment by Techeryy** [comment]
> One thing you can try setting is `GOMAXPROCS=1` in the container env, so Coder doesn't lock up the whole server. This will force Coder to only use 1 core. To help figure out where the CPU is going, you can add `CODER_PPROF_ENABLE=true` and forward port `6060` to the host.
> 
> When you see the CPU pinning, run `curl -o cpu-trace.pprof http://localhost:6060/debug/pprof/trace?seconds=10` to grab a CPU profile. If you could send it back to us via github or email (`<first name>@coder.com` for any of us), that would be great.

Sorry for the delay in my response, after setting the environment variables as instructed above, PPROF shows that it's enabled & configurated correctly on the coder deployment tab however when I run the curl command to grab the CPU profile I get one of two responses:

1. `curl: (52) Empty reply from server`
2. `curl: (56) Recv failure: Connection reset by peer`

When the behaviour occurred, the CPU continued to operate at 100% utilization, despite being configured to only use a single core. Despite this I could still execute the `curl` command, it took a considerable amount of time to receive a response. As previously mentioned, the response from running the command gave one of the statements mentioned above.

This is a paste of my docker compose file just in case I made a mistake: https://pastebin.com/yzWGdYGL

**Comment by johnstcn** [comment]
Whoops, I think you also need to set `CODER_PPROF_ADDRESS: "0.0.0.0:6060"`, otherwise it just listens on `127.0.0.1:6060` inside the container.

No problem though, you should still be able to curl the endpoint inside the container and then copy the file outside the container:

```
docker exec -it coder_coder_1 curl -fsSL -o trace.out 'http://localhost:6060/debug/pprof/trace?seconds=1'
docker cp coder_coder_1:/home/coder/trace.out trace.out
```

**Comment by johnstcn** [comment]
Could you also send us the output of `top -n1` on your VPS?

**Comment by Techeryy** [comment]
Sorry I have taken a while with my response, I have moved onto to some different programming projects in the last week, away from web design, and was waiting to experience the issue again to generate you a report. However I haven't experienced the behaviour since, despite not even changing Coder versions. I would appreciate it if this issue report could be kept open for 2-3 weeks to see if I have any issues & if I do I will run the command mentioned above to get you a detailed overview on what's happening with my system.

I have also linked the output of running `top -n1` on my VPS:
https://pastebin.com/yUMy5JWq

**Comment by Techeryy** [comment]
Sorry for the excessively long response time, I believe the issue I was encountering was due to the specification of my VPS. I have now deployed on a more powerful server & had no issues with the version in question or any later versions. It is likely that my server was only just coping with the load Coder was putting on the server & upon updating to V2.7.2, a change made in the update must have made the load slightly to much for my hardware and hence caused it to freeze at random intervals.
Thank You for looking into this for me, I now use Coder on a daily basis & really appreciate the work you do!
